### ISSUE TITLE
Add `OnceLock::wait`

### ISSUE THREAD

**Issue by tgross35** [rust-lang/libs-team#405]
# Proposal

## Problem statement

`OnceCell` is useful for synchronizing write-once data, but it does not have any blocking methods that allow other threads to wait for this data.

## Motivating examples or use cases

Sometimes it can be useful to allow the end of computation to return with a value across thread boundaries, when joining may not be acceptable. For example, [rayon::spawn](https://docs.rs/rayon/latest/rayon/fn.spawn.html) does not have a return value that allows for easy returning. Other applications are nested spawning, or cases where a worker thread should return a value (and unblock the parent thread) then continue going without joining.

The example from [once_cell](https://docs.rs/once_cell/latest/once_cell/sync/struct.OnceCell.html#method.wait) shows the basic structure:

```rust
use once_cell::sync::OnceCell;

let mut cell = std::sync::Arc::new(OnceCell::new());
let t = std::thread::spawn({
    let cell = std::sync::Arc::clone(&cell);
    move || cell.set(92).unwrap()
});

// Returns immediately, but might return None.
let _value_or_none = cell.get();

// Will return 92, but might block until the other thread does `.set`.
let value: &u32 = cell.wait();
assert_eq!(*value, 92);
t.join().unwrap();
```

## Solution sketch

Add the following:

```rust
impl OnceLock {
    /// Block the current thread until the value is set, then return it
    fn wait(&self) -> &T;
}
```

## Alternatives

Approximating this behavior with other synchronization primitives, e.g. `Barrier` to wait for the thread and `Mutex` to store the return value, or just manually parking (which this method is a thin wrapper around).

## Links and related work

- https://docs.rs/once_cell/latest/once_cell/sync/struct.OnceCell.html#method.wait

## What happens now?

This issue contains an API change proposal (or ACP) and is part of the libs-api team [feature lifecycle]. Once this issue is filed, the libs-api team will review open proposals as capability becomes available. Current response times do not have a clear estimate, but may be up to several months.

[feature lifecycle]: https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html

## Possible responses

The libs team may respond in various different ways. First, the team will consider the *problem* (this doesn't require any concrete solution or alternatives to have been proposed):

- We think this problem seems worth solving, and the standard library might be the right place to solve it.
- We think that this probably doesn't belong in the standard library.

Second, if there's a concrete solution:

- We think this specific solution looks roughly right, approved, you or someone else should implement this. (Further review will still happen on the subsequent implementation PR.)
- We're not sure this is the right solution, and the alternatives or other materials don't give us enough information to be sure about that. Here are some questions we have that aren't answered, or rough ideas about alternatives we'd want to see discussed.

**Comment by tgross35** [comment]
Hm, the `once_cell` implementation does this by parking with a custom key https://github.com/matklad/once_cell/blob/c48d3c2c01de926228aea2ac1d03672b4ce160c1/src/imp_pl.rs#L80-L92 (cc @matklad). We don't currently have a way to block on `Once`,  but maybe we could add one internally.

**Comment by matklad** [comment]
@tgross35 you probably want to look at the std impl, it should be closer to what std needs to do: https://github.com/matklad/once_cell/blob/c48d3c2c01de926228aea2ac1d03672b4ce160c1/src/imp_std.rs#L89

**Comment by m-ou-se** [comment]
We discussed this in the libs-api meeting just now. We'd like to see this API added, and also on `Once`. We like the name `wait`, as it's very obvious what to wait for.

Feel free to [open a tracking issue](https://github.com/rust-lang/rust/issues/new?assignees=&labels=C-tracking-issue%2C+T-libs&template=library_tracking_issue.md&title=Tracking+Issue+for+XXX) and open a PR to rust-lang/rust to add it as an unstable feature.

We did not discuss the potential implementation in detail.

It might be worth exploring what it would take to also have a .await'able version in the future.
